### ISSUE TITLE
I am getting corrupted text for Unicode escaped strings when in a UTF8 buffer

### ISSUE THREAD

**Issue by jango67** [Tencent/rapidjson#1857]
I am new to rapidjson but have been using a wrapped library for years. I recently got involved with a web project in japan and this is where I ran into an issue. My specs.

I believe I am using rapidjson 1.0.2
Visual Studio 2010 SP1
C++
On windows.

I have a UTF8 buffer with this:
"id":"08fe8e01-a3ac-4b1c-b305-7d8888888888",
"deletedDateTime":null,"classification":null,"createdDateTime":"2021-03-11T17:44:56Z",
"creationOptions":[],"description":null,"displayName":"\u5947\u5999\u306a",
"expirationDateTime":null,"groupTypes":[],"isAssignableToRole":null,"mail":"weirdtest@jpptesto365.onmicrosoft.com",
"mailEnabled":true,"mailNickname":"weirdtest",
"membershipRule":null,"membershipRuleProcessingState":null,"onPremisesDomainName":null,"onPremisesLastSyncDateTime":null,"onPremisesNetBiosName":null,"onPremisesSamAccountName":null,"onPremisesSecurityIdentifier":null,"onPremisesSyncEnabled":null,"preferredDataLocation":null,"preferredLanguage":null,"proxyAddresses":["SMTP:weirdtest@jpptesto365.onmicrosoft.com"],"renewedDateTime":"2021-03-11T17:44:56Z",
"resourceBehaviorOptions":[],"resourceProvisioningOptions":[],"securityEnabled":false,"securityIdentifier":"S-1-12-1-150900225-1260168108-2273117619-3098968976",
"theme":null,"visibility":null,"onPremisesProvisioningErrors":[]},

This just happens to be a group with a japanese name. 
When this is translated by document::parse I get some funky characters. I am not sure if the person that originally wrote the wrapper ever tested with asian character sets.  This is what I get for my name after it has been parsed:

å¥.å¦™ã.ª

what I should get is 

奇妙な

I am not sure how to proceed. I have debugged this implementation and found that is pretty simplistic and seems to follow recommendations I have found online for using rapidjson.  My research keeps telling me that the escaped unicode characters will be converted to UTF16 while the rest of document is using UTF8. Based on documentation it seems that this should work but after debugging it I am not sure if it should.  This is a high profile project and I have been asked to start testing other JSON parsing libraries to see if they have same issue. Since I really do not want to spend time looking at something else I figured I would post this question and see if I can get a solution. Any help figuring this out would be very much appreciated .

**Comment by miloyip** [comment]
The default is parsing UTF-8 source JSON into UTF-8 memory reprsentation:

~~~cpp
#include "rapidjson/document.h"

int main() {
    rapidjson::Document d;
    d.Parse("\"\u5947\u5999\u306a\"");
    puts(d.GetString());
}
~~~

~~~
$ g++ -I /github/rapidjson/include a.cpp && ./a.out 
奇妙な
~~~

On Mac, Linux, UTF-8 is the usual way in memory representation.
On Windows, most native API requires UTF-16.

You may use rapidjson's encoding conversion. But if your wrapping library does not support that, you need to do the conversion after `GetString()` calls. For example, Win32 provides `MultiByteToWideChar()` for that.

**Comment by jango67** [comment]
We do have a thin wrapper on rapidjson but when I debugged it the string I am looking at is for a rapidjson::Value. At this point the wrapping code is not involved. I was concerned that the version could be an issue or maybe because my app is 32bit app. I am not sure why it does not seem to convert the escaped unicode sequence correctly. I will keep why you have offered in mind as a track down the root cause of the issue. When looking at the rapidjson::value the data.ss and data.payload. payload flags are 7173. I not sure what flags are yet but hope to dig in a bit more tomorrow. Customer for project is less than happy so I am in the hot seat for sure. Thanks for some info, if you think of anything else I welcome any info that could be helpful!

**Comment by miloyip** [comment]
It is not related to 32-bit or 64-bit Windows. It is just the encoding issue.
If your Windows application needs UTF-16 for display or process, just use `MultiByteToWideChar()` to convert from `char *` UTF-8 string to `wchar_t *` UTF-16 string.

**Comment by jango67** [comment]
I have tried several options and still can't get rapidjson to decode stuff correctly. This seems like it should work. 

	    rapidjson::Document d;
		rapidjson::Value*	pJsonVal;
		rapidjson::Value*	pJsonVal2;
		char *zaq = "{\"value\":[{\"description\":null,\"displayName\":\"\\u5947\\u5999\\u306a\"},{\"description\":\"This is the default group for everyone in the network\",\"displayName\":\"All Company\"},{\"description\":null,\"displayName\":\"alanaGroup\"}]}";
//		char zaq[4096];
//		memset( zaq,0,4096);
//		WideCharToMultiByte(CP_UTF8, 0, _T("{\"value\":[{\"description\":null,\"displayName\":\"\\u5947\\u5999\\u306a\"},{\"description\":\"This is the default group for everyone in the network\",\"displayName\":\"All Company\"},{\"description\":null,\"displayName\":\"alanaGroup\"}]}"), -1, zaq, sizeof(zaq), NULL, NULL);

//	    d.Parse("\"\u5947\u5999\u306a\"");
		d.Parse(zaq);
		if( d.HasMember("value") )
		{
			pJsonVal = &(d["value"]);
		}
		pJsonVal2 = &(*pJsonVal)[0];
		if( pJsonVal2->HasMember( "displayName" ) )
		{
			pJsonVal = &(*pJsonVal2)["displayName"];
		}
	    puts(pJsonVal->GetString());

After reading what seems to be millions of lines of documentation on rapidjson it seems like this should work. I am off to try new version to see if that fixes anything.

**Comment by jango67** [comment]
latest version same issue. I was hoping not to debug the json lib itself but definitely not working correctly so I have not choice at this point.

**Comment by jango67** [comment]
After some additional debugging and research it appears that rapidjson is converting "\u5947\u5999\u306a" to these characters e5 a5 87 e5 a6 99 e3 81 aa. I tracked this down by debugging rapidjson (nice code to debug, no something you find in third party lib often) and cross referencing that with information from here https://en.wikipedia.org/wiki/UTF-8 (rapidjson also had this reference in there code) I was able to determine what I am getting out of rapidjson is correct. I appreciate all the help and keep up the good work on rapidjson. I know when I recommended it years ago it was the right choice. Thanks again, this not an issue in RapidJson.
